### PR TITLE
Upgrade SDK for beta-4 compatibility

### PR DESCRIPTION
**Pull request by sdankel**
Closes https://github.com/FuelLabs/sway-playground/issues/33

To test, use the latest wallet: https://fuels-wallet-m1jd4682z-fuel-labs.vercel.app/app/fuel-wallet.zip

Run the front end and backend locally per the readme.

Summary of the changes:
- using a type map to determine the types of the ABI methods (required with the new version of fuels)
- changes to how we transform the data for the contract interface inputs (required with the new version of fuels)
- replacing `useFuel` and `useProvider` with the hooks from the newly available`@fuel-wallet/react`
- actually reading the compiler output in `storage_slots.json` and passing that along to the SDK during deployment (missing this was causing a generic revert)
- Adding beta-4 to the toolchain dropdown, docker image, and making it the default

### CODE REVIEW

**Comment by sdankel** [comment]
Blocked by https://github.com/FuelLabs/fuels-ts/issues/1245

**Comment by sdankel** [comment]
We should wait for https://github.com/FuelLabs/fuels-npm-packs/issues/20 to use a more robust Fuel Provider. Currently the issue is that the way we're fetching the provider (to get the networkUrl) is returning null for beta-4. This is because we're not listening for events that update the provider.

**Comment by sdankel** [comment]
It's all working now, except that the beta-4 network seems to be rejecting function calls. Even for the simplest example contracts, the response from the network is `Error: The script reverted with reason Unknown`. It was responding correctly yesterday, so this needs more investigation. The SDK version hasn't changed since then.

File: app/package.json
Comment (cold-briu) [inline app/package.json:1]: - fuel-wallet packages current version is 0.13.0 
- fuels (fuel-ts) current version is 0.59

File: deployment/Dockerfile
Comment (cold-briu) [inline deployment/Dockerfile:1]: I would remove versions below beta-3 since are not supported

**Review by cold-briu — COMMENTED** [review]
Please bump fuels (fuel-ts) and fuel-wallet versions

File: app/src/features/interact/components/FunctionInterface.tsx [hunk 1]
```diff
@@ -22,55 +31,73 @@ export interface FunctionInterfaceProps {
...
-  }, [functionFragment]);
+    const outputTypeId = functionFragment?.jsonFn.output?.type;
+    if (outputTypeId !== undefined) {
+      const sdkType = typeMap.get(outputTypeId);
+      return sdkType ? getTypeInfo(sdkType, typeMap).literal : undefined;
+    }
+    
+  }, [functionFragment?.jsonFn.output, typeMap]);
+
+  // const outputType = undefined;
```
Comment (eureka-cpu) [inline app/src/features/interact/components/FunctionInterface.tsx:91]: can this be removed?

File: app/src/features/interact/components/FunctionInterface.tsx [hunk 2]
```diff
@@ -22,55 +31,73 @@ export interface FunctionInterfaceProps {
...
     () =>
-      functionFragment?.inputs.map((input) =>
-        toInputInstance(input as SdkParamType)
+      functionFragment?.jsonFn.inputs.map((input) =>
+        toInputInstance(input.type, input.name)
       ) ?? [],
-    [functionFragment?.inputs, toInputInstance]
+    [functionFragment?.jsonFn.inputs, toInputInstance]
   );
+  // const inputInstances: InputInstance[] = [];
```
Comment (eureka-cpu) [inline app/src/features/interact/components/FunctionInterface.tsx:100]: here too

File: deployment/Dockerfile
Comment (sdankel) [reply deployment/Dockerfile:1]: I thought about this too - technically those versions are for the forc toolchain, which are supported. What isn't supported is older versions of the wallet. Users might still find value in being able to compile sway code with older toolchain versions, even if they're not able to deploy them. If it's too confusing I can go ahead and remove beta-1, beta-2 and beta-3 from the dropdown.

**Comment by sarahschwartz** [comment]
Can we update the docs link to `https://docs.fuel.network/docs/sway/`? Also, I tried deploying a contract to beta-4 and got the error `Deployment failed: "undefined" is not valid JSON`.

**Comment by sdankel** [comment]
> Can we update the docs link to `https://docs.fuel.network/docs/sway/`? Also, I tried deploying a contract to beta-4 and got the error `Deployment failed: "undefined" is not valid JSON`.

You will need to run the backend locally (`cargo run` from the root directory) and point the frontend at it by making the following change in `app/src/features/editor/hooks/useCompile.tsx`:

```diff
-    const server_uri = 'https://api.sway-playground.org/compile';
-    // const server_uri = 'http://0.0.0.0:8080/compile';
+    // const server_uri = 'https://api.sway-playground.org/compile';
+    const server_uri = 'http://0.0.0.0:8080/compile';
```

**Comment by luizstacio** [comment]
@sdankel for some reason, the deployContract timeouts I'm using version 0.13 of the wallet. https://chrome.google.com/webstore/detail/fuel-wallet-development/hcgmehahnlbhpilepakbdinkhhaackmc Debugging the messages from the Wallet the wallet is responding with the transaction ID as expected.

File: README.md [hunk 3]
```diff
@@ -70,7 +77,16 @@ cd app
...
+This will open http://localhost:3000 in your browser. By default, it will use the production backend endpoint. 
+
+To test against the backend running locally, make this change in `app/src/features/editor/hooks/useCompile.tsx`:
+
+```diff
+-    const server_uri = 'https://api.sway-playground.org/compile';
+-    // const server_uri = 'http://0.0.0.0:8080/compile';
++    // const server_uri = 'https://api.sway-playground.org/compile';
++    const server_uri = 'http://0.0.0.0:8080/compile';
+```
```
Comment (luizstacio) [inline README.md:89]: I think would be cool to add a check if by the `NODE_ENV` and on `development` just use the local compiler.

**Comment by luizstacio** [comment]
> @sdankel for some reason, the deployContract timeouts I'm using version 0.13 of the wallet. https://chrome.google.com/webstore/detail/fuel-wallet-development/hcgmehahnlbhpilepakbdinkhhaackmc Debugging the messages from the Wallet the wallet is responding with the transaction ID as expected.

I found that the issue is on the chain and not on the code itself. Everything should work fine once it's resolved.

**Comment by luizstacio** [comment]
@sdankel on my side I don't see any many things that could be changed right now.  Just left a small comment to improve running locally.

On the useWallet and useDeploy, I think we can have some improvements to not need the isDisable param, but it looks to not be related to this PR.

File: README.md (hunk 3 above)
Comment (sdankel) [reply README.md:89]: Yes, I have a TODO for that, I'll do it in a separate PR

**Review by luizstacio — APPROVED** [review]

**Pull request merged**